### PR TITLE
Iowa state income tax

### LINKED ISSUE #2044: Iowa state income tax

### PR DESCRIPTION
**Pull request by martinholmer**
Fixes #2044.

When the DRAFT label was removed from this pull request on 2023-05-06, there were 19,158 IA income tax differences between the code in this PR and TAXSIM35 as patched through 2023-05-06 for the `p21` sample of 100,000 randomly-generated integration tests.  The largest of these `p21` differences was $47.96.   The results in issue #993 show that these are better validation results than for either UT or NY, both of which have already been merged into the master branch.

The variable `ia_income_tax` cannot be added to the list of state income taxes in the `state_income_tax.py` file because of a circular logic error.  Here is an example of what happens when `ia_income_tax` is included in the list:
```
policyengine_core.errors.cycle_error.CycleError: Circular definition detected on formula income_tax_before_refundable_credits@2021. The circle is:

Normal computation tree:
  income_tax@2021 (branch default)

Cycle start:
  >> iitax@2021 (branch default)
  >> income_tax_before_refundable_credits@2021 (branch default)
  >> income_tax_before_credits@2021 (branch default)
  >> income_tax_main_rates@2021 (branch default)
  >> taxable_income@2021 (branch default)
  >> taxable_income_deductions@2021 (branch default)
  >> taxable_income_deductions_if_itemizing@2021 (branch default)
  >> salt_deduction@2021 (branch default)
  >> state_and_local_sales_or_income_tax@2021 (branch default)
  >> state_income_tax@2021 (branch default)
  >> ia_income_tax@2021 (branch default)
  >> ia_income_tax_before_refundable_credits@2021 (branch default)
  >> ia_income_tax_before_credits@2021 (branch default)
  >> ia_files_separately@2021 (branch default)
  >> ia_income_tax_indiv@2021 (branch default)
  >> ia_base_tax_indiv@2021 (branch default)
  >> ia_regular_tax_indiv@2021 (branch default)
  >> ia_taxable_income_indiv@2021 (branch default)
  >> ia_fedtax_deduction@2021 (branch default)
  >> income_tax_before_refundable_credits@2021 (branch default)
```

### Summary
:sparkles::page_facing_up::construction:

This pull request adds Iowa state income tax to the policy engine. It creates new parameter, variable, and test files for the components of the Iowa income tax system, such as net income, deductions, rates, and basic tax. It also updates the changelog file to reflect the minor version bump. Some of the formulas and tests are not yet implemented and will be completed in future commits.

> _`Iowa` tax code_
> _New parameters and tests_
> _Autumn leaves fall fast_

### Walkthrough
*  Add Iowa state income tax to the policy engine ([link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-5885042762833a4aef51405a4036067e4e6aefa7a5091a2987d1d9d1fc2647e8R1-R4))
*  Create new parameter files for the Iowa income tax components, such as rates, brackets, deductions, additions, and subtractions ([link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-100be9e54bd8503f6268090c4e89beece5b05fcbaaafcb9d6408e8d1ee111f85R1-R19), [link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-52f551ce170cc4d64d3950d034a139f88eae4afc0992939fa2805202d7d65b02R1-R19), [link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-1fb4b6cf4bfddf079572e74e1a09ecc63d15c498fbce0db1bdb67432413af96aR1-R19), [link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-4c7e9e1d9cd9a50e975ae3802260b214c57206fb3f7ba4b25e661172f1ad2510R1-R21), [link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-801744d03bb19100d457c9fda3afaa1ab4eb9ba4f64c6b46f2474ed03373ab59R1-R58))
*  Create new variable files for the Iowa income tax calculations, such as net income, taxable income, basic tax, and deductions ([link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-165ce0a35c9fc11101f40094d1b751fd347ad840b89c775697e9c3a14e331326R1-R28), [link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-63d8464d46e6a6a34886f154dcfd0efe2083c41afd7f24423715f739afe424d9R1-R19), [link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-b8e0fbe370f73044c4f93ac058d84a6bd7a1ccce2eaba6478b06c9fc73192c86R1-R26), [link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-ecda8539612014649096b46039966e5b32acbd3064ee3aa805056ea72376a651R1-R21), [link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-0298b25a2dde75f7e050a1f93856916aad6bd7c61e8520915c51f18419952a7dR1-R20), [link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-3d52bf78936676686bbefb9f8d461e7348246beb38efcdebb4f893c050b9ceceR1-R21), [link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-8601a003d901e6698350e37b82818117f94e7293f7506cd96fe0a827911103dfR1-R20), [link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-2edda1c8f8379b2bb18c45e797de7cb926e408eef05a811606ffbaa3327a99e4R1-R22), [link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-96983f64912994549bb286978d4cc273b02a82ca5fc59e586b6ef24c97ea495cR1-R34), [link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-c0bbb173206ffd7b4bd104c08723032bf3b6d102bbf2c83ca62e756e130a5c18R1-R33))
*  Implement formulas for some of the variables, such as `ia_basic_deduction`, `ia_qbi_deduction`, `ia_basic_tax`, and `ia_taxable_income` ([link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-165ce0a35c9fc11101f40094d1b751fd347ad840b89c775697e9c3a14e331326R1-R28), [link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-ecda8539612014649096b46039966e5b32acbd3064ee3aa805056ea72376a651R1-R21), [link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-3d52bf78936676686bbefb9f8d461e7348246beb38efcdebb4f893c050b9ceceR1-R21), [link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-2edda1c8f8379b2bb18c45e797de7cb926e408eef05a811606ffbaa3327a99e4R1-R22))
*  Create new test files for some of the variables, such as `ia_basic_deduction`, `ia_qbi_deduction`, `ia_basic_tax`, and `ia_taxable_income` ([link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-5179f64ac572b881f51b85f63e2124cf91f23287d04e3704820905977227c731R1-R19), [link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-a846fba50f3f4e7608035743c137a7220dd8d922c236b50381bfefa1e99fb1d8R1-R17), [link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-94f8801c1fc1296a3572639c7de4265bd6310ec16e122465956198355fc84546R1-R35), [link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-6498d025813f6104b640ace32c2bcd167a4c800457df3bddef52083af67c4018R1-R23))
*  Specify unit tests for the variables, using input and output values from the Iowa Department of Revenue or other sources ([link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-5179f64ac572b881f51b85f63e2124cf91f23287d04e3704820905977227c731R1-R19), [link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-a846fba50f3f4e7608035743c137a7220dd8d922c236b50381bfefa1e99fb1d8R1-R17), [link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-94f8801c1fc1296a3572639c7de4265bd6310ec16e122465956198355fc84546R1-R35), [link](https://github.com/PolicyEngine/policyengine-us/pull/2138/files?diff=unified&w=0#diff-6498d025813f6104b640ace32c2bcd167a4c800457df3bddef52083af67c4018R1-R23))

### CODE REVIEW

File: policyengine_us/variables/gov/states/ia/tax/income/deductions/ia_itemized_deductions.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    defined_for = StateCode.IA
+
+
+#    def formula(tax_unit, period, parameters):
+#        """
+#        FROM THE 2021 INSTRUCTIONS (PAGE 46):
+#          The $10,000 federal cap on the itemized deduction for state and
+#        local taxes does not apply for Iowa purposes. Taxpayers may still
+#        deduct eligible state and local taxes paid, independent of the
+#        federal dollar limitation.
```
Comment (nikhilwoodruff) [inline policyengine_us/variables/gov/states/ia/tax/income/deductions/ia_itemized_deductions.py:25]: Sorry what does this mean? Not sure I follow the the SALT deduction in the context of State taxes here.

**Review by nikhilwoodruff — COMMENTED** [review]
Thanks Martin- looks great so far.

**Review by MaxGhenis — COMMENTED** [review]
Thanks Martin - looks great, just a couple small things.

File: policyengine_us/tests/policy/baseline/gov/states/ia/tax/income/ia_fedtax_deduction.yaml [hunk 2]
```diff
@@ -0,0 +1,28 @@
+- name: IA federal income tax deduction unit test 1
```
[the remark this answers is not in the document]
Comment (martinholmer) [reply policyengine_us/tests/policy/baseline/gov/states/ia/tax/income/ia_fedtax_deduction.yaml:1]: Done.   
But you ought to require complete code coverage for **all** PRs, not just mine.

File: policyengine_us/variables/gov/states/ia/tax/income/deductions/ia_standard_deduction_indiv.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        "https://tax.iowa.gov/sites/default/files/2023-01/2021%20Expanded%20Instructions_010323.pdf"
+        "https://tax.iowa.gov/sites/default/files/2023-01/2022IA1040%2841001%29.pdf"
+        "https://tax.iowa.gov/sites/default/files/2023-03/2022%20Expanded%20Instructions_022023.pdf"
+    )
+    defined_for = StateCode.IA
+
+    def formula(person, period, parameters):
+        us_filing_status = person.tax_unit("filing_status", period)
+        fsvals = us_filing_status.possible_values
+        filing_status = select(
```
[the remark this answers is not in the document]
Comment (martinholmer) [reply policyengine_us/variables/gov/states/ia/tax/income/deductions/ia_standard_deduction_indiv.py:21]: Excellent suggested code simplification.  Thanks.
I've implemented this change in `ia_standard_deduction_indiv.py` and in the
`ia_amt_indiv.py` module.   
But I think making it a separate variable seems more trouble than its worth.

[the remark this answers is not in the document]
Comment (martinholmer) [reply policyengine_us/variables/gov/states/ia/tax/income/deductions/ia_standard_deduction_indiv.py:21]: After implementing your suggested code simplification (which I now remember trying initially) produces this error:
```
_________________________________ test session _________________________________
//Users/mrh/anaconda3/envs/policyengine-us/lib/python3.9/site-packages/_pytest/runner.py:244: in from_call
    result = func()
//Users/mrh/anaconda3/envs/policyengine-us/lib/python3.9/site-packages/_pytest/runner.py:217: in <lambda>
    lambda: ihook(item=item, **kwds), when=when, reraise=reraise
//Users/mrh/anaconda3/envs/policyengine-us/lib/python3.9/site-packages/pluggy/hooks.py:286: in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
//Users/mrh/anaconda3/envs/policyengine-us/lib/python3.9/site-packages/pluggy/manager.py:93: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
//Users/mrh/anaconda3/envs/policyengine-us/lib/python3.9/site-packages/pluggy/manager.py:84: in <lambda>
    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
//Users/mrh/anaconda3/envs/policyengine-us/lib/python3.9/site-packages/pluggy/callers.py:208: in _multicall
    return outcome.get_result()
//Users/mrh/anaconda3/envs/policyengine-us/lib/python3.9/site-packages/pluggy/callers.py:80: in get_result
    raise ex[1].with_traceback(ex[2])
//Users/mrh/anaconda3/envs/policyengine-us/lib/python3.9/site-packages/pluggy/callers.py:187: in _multicall
    res = hook_impl.function(*args)
//Users/mrh/anaconda3/envs/policyengine-us/lib/python3.9/site-packages/_pytest/runner.py:143: in pytest_runtest_call
    raise e
//Users/mrh/anaconda3/envs/policyengine-us/lib/python3.9/site-packages/_pytest/runner.py:135: in pytest_runtest_call
    item.runtest()
//Users/mrh/anaconda3/envs/policyengine-us/lib/python3.9/site-packages/policyengine_core/tools/test_runner.py:260: in runtest
    self.check_output()
//Users/mrh/anaconda3/envs/policyengine-us/lib/python3.9/site-packages/policyengine_core/tools/test_runner.py:289: in check_output
    self.check_variable(
//Users/mrh/anaconda3/envs/policyengine-us/lib/python3.9/site-packages/policyengine_core/tools/test_runner.py:332: in check_variable
    actual_value = self.simulation.calculate(variable_name, period)
//Users/mrh/anaconda3/envs/policyengine-us/lib/python3.9/site-packages/policyengine_core/simulations/simulation.py:330: in calculate
    result = self._calculate(variable_name, period)
//Users/mrh/anaconda3/envs/policyengine-us/lib/python3.9/site-packages/policyengine_core/simulations/simulation.py:491: in _calculate
    array = self._run_formula(variable, population, period)
//Users/mrh/anaconda3/envs/policyengine-us/lib/python3.9/site-packages/policyengine_core/simulations/simulation.py:734: in _run_formula
    array = formula(population, period, parameters_at)
policyengine_us/variables/gov/states/ia/tax/income/deductions/ia_standard_deduction_indiv.py:29: in formula
    return (is_head | is_spouse) * p.deductions.standard[filing_status]
//Users/mrh/anaconda3/envs/policyengine-us/lib/python3.9/site-packages/policyengine_core/parameters/parameter_node_at_instant.py:58: in __getitem__
    return parameters.VectorialParameterNodeAtInstant.build_from_node(
//Users/mrh/anaconda3/envs/policyengine-us/lib/python3.9/site-packages/policyengine_core/parameters/vectorial_parameter_node_at_instant.py:216: in __getitem__
    self.vector[key[0]], numpy.nan
//Users/mrh/anaconda3/envs/policyengine-us/lib/python3.9/site-packages/numpy/core/records.py:510: in __getitem__
    obj = super().__getitem__(indx)
E   ValueError: no field of name 0
```
Not very informative to me, but my guess is that somehow `us_filing_status` is not represented in a way that is compatible with the ENUM type for filing status.   The only way I could avoid this error was to use the verbose syntax you objected to.

I suggest you merge the PR and then ask somebody to conduct research into why your suggestion does not work.

**Pull request merged**